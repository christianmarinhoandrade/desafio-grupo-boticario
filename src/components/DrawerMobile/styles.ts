import styled from 'styled-components'

export const TitleContent = styled.div`
  display: flex;
  justify-content: center;
`
