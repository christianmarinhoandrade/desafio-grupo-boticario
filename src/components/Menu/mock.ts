export default [
  {
    title: 'Cadastro Revendedor',
    path: 'revendedor/inclusao',
    name: 'revendedor-inclusao',
    'data-testid': 'menu-item-id'
  },
  {
    title: 'Cadastro Compras',
    path: 'compras/inclusao',
    name: 'compras-inclusao',
    'data-testid': 'menu-item-id'
  }
]
