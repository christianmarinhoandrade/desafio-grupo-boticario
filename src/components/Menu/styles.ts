import styled from 'styled-components'

import { Menu as MenuAntd } from 'antd'

export const Menu = styled(MenuAntd)`
  width: 100%;
`
