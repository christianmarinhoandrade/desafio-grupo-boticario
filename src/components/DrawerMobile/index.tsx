import { Drawer, Image } from 'antd'
import { TitleContent } from './styles'

import grupoBoticario from '@/img/logo-grupo-boticario.png'

export type DrawerMobileProps = {
  children: JSX.Element
  setVisible?: (visible: boolean) => void
  visible: boolean
}
const DrawerMobile = ({ children, setVisible, visible }: DrawerMobileProps) => {
  return (
    <Drawer
      title={
        <TitleContent>
          <Image
            src={grupoBoticario}
            width={100}
            preview={false}
            alt="Grupo BoticÃ¡rio"
          />
        </TitleContent>
      }
      placement="left"
      closable={false}
      onClose={() => {
        if (setVisible) setVisible(false)
      }}
      visible={visible}
    >
      {children}
    </Drawer>
  )
}

export default DrawerMobile
