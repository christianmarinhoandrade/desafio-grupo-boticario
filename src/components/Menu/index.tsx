import * as S from './styles'

type ItemMenuProps = {
  path: string
  title: string
  name: string
}

type MenuProps = {
  menu: ItemMenuProps[]
  mode?: 'horizontal' | 'inline'
  mobile?: boolean
  signOut?: () => void
  onClick?: (route: string) => void
}

function Menu({
  menu,
  mode = 'horizontal',
  signOut,
  onClick,
  mobile = false,
  ...props
}: MenuProps) {
  const renderMenu = (menus: ItemMenuProps[]) => {
    return menus.map(({ path, title, name, ...props }: ItemMenuProps) => (
      <S.Menu.Item
        onClick={() => {
          if (path && onClick) onClick(path)
        }}
        key={name}
        {...props}
      >
        {title}
      </S.Menu.Item>
    ))
  }

  return (
    <S.Menu mode={mode} {...props}>
      {renderMenu(menu)}
      {mobile && (
        <S.Menu.Item
          key="sair"
          onClick={() => {
            if (signOut) signOut()
          }}
        >
          Sair
        </S.Menu.Item>
      )}
    </S.Menu>
  )
}

export default Menu
