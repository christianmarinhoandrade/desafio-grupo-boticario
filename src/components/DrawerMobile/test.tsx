import { render, waitFor } from '@testing-library/react'

import DrawerMobile from '.'
import Menu from '~/components/Menu'
import mockItems from '~/components/Menu/mock'

describe('<DrawerMobile />', () => {
  const menu = mockItems
  it('should render drawer mobile default', async () => {
    const { container } = render(
      <DrawerMobile visible={true}>
        <Menu menu={menu} data-testid={'menu-id'} />
      </DrawerMobile>
    )

    await waitFor(() => {
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
