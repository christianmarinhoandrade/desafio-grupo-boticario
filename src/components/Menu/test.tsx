import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Menu from '.'
import mockItems from './mock'

describe('<Menu />', () => {
  const menu = mockItems

  it('should render menu default', async () => {
    const { container } = render(<Menu menu={menu} data-testid={'menu-id'} />)
    await waitFor(() => {
      expect(screen.getByTestId('menu-id')).toBeInTheDocument()
      expect(screen.queryByText('Sair')).not.toBeInTheDocument()
      expect(screen.getByTestId('menu-id')).toHaveClass('ant-menu-horizontal')
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  it('should render menu with mode inline', async () => {
    render(<Menu menu={menu} data-testid={'menu-id'} mode="inline" />)
    await waitFor(() => {
      expect(screen.getByTestId('menu-id')).toBeInTheDocument()
      expect(screen.getByTestId('menu-id')).toHaveClass('ant-menu-inline')
    })
  })

  it('should render menu with mobile true', async () => {
    render(<Menu menu={menu} data-testid={'menu-id'} mobile={true} />)
    await waitFor(() => {
      expect(screen.getByTestId('menu-id')).toBeInTheDocument()
      expect(screen.getByText('Sair')).toBeInTheDocument()
    })
  })

  it('should render menu with a two items', async () => {
    render(<Menu menu={menu} data-testid={'menu-id'} mobile={true} />)
    await waitFor(() => {
      expect(screen.getAllByTestId('menu-item-id')).toHaveLength(2)
    })
  })

  it('should render menu to call the methods signOut and onClick if clicked', async () => {
    const onClick = jest.fn()
    const signOut = jest.fn()
    render(
      <Menu
        menu={menu}
        data-testid={'menu-id'}
        onClick={onClick}
        signOut={signOut}
        mobile={true}
      />
    )

    const menuItem = screen.getByText(menu[0].title)
    const menuItemExit = screen.getByText('Sair')

    userEvent.click(menuItemExit)
    userEvent.click(menuItem)
    await waitFor(() => {
      expect(onClick).toHaveBeenCalledTimes(1)
      expect(signOut).toHaveBeenCalledTimes(1)
    })
  })
})
