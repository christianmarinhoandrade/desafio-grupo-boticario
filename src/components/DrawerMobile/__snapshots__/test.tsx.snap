// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawerMobile /> should render drawer mobile default 1`] = `null`
